### ISSUE TITLE
Fairness issue after 100% slash

### ISSUE THREAD

**Issue by yazzaoui** [autonity/autonity#819]
_Originally posted by @lorenzo-dev1 in https://github.com/autonity/autonity/issues/807#issuecomment-1651083480_

Scenario:

Supposing we have a registered validator with no self-bonded stake. Then delegator1 bonds 10 NTN to it.

At the end of the epoch we will have:
```
_validator.liquidSupply = 10
_validator.bondedStake = 10
```

Then delegator1 unbonds. The LNTN gets locked, then at the end of the epoch we will have:
```
_validator.liquidSupply = 0
 _validator.bondedStake = 0
_validator.unbondingStake = 10
_validator.unbondingShares = 10
_unbonding.unbondingShare = 10
```

Then before the unbonding period expires, the validator gets slashed due a past offence with critical severity (100% slash) and all of the unbondingStake gets slashed:
```
_validator.unbondingStake = 0
```

So delegator1 should receive 0 NTN when the unbonding period expires.

But let's assume that, always before the unbonding period expires, delegator2 bonds 10 NTN:
```
_validator.liquidSupply = 10
_validator.bondedStake = 10
```

And unbonds 10 NTN:
```
_validator.liquidSupply = 0
 _validator.bondedStake = 0
_validator.unbondingStake = 10
_validator.unbondingShares = 20
_unbonding.unbondingShare = 10
```

Now when the two unbonding periods expire, what is going to happen is that both delegator1 and delegator2 will take the hit from the slashing, instead of only delegator1.

~Agreed upon solution: Enforce minimum of `unbondingPeriod` between bond and unbond operations.~

TBD The issue is also valid for the bonding NTN -> LNTN operation

**Comment by yazzaoui** [comment]
Considering to introduce the special state `killed` if slashed at 100% but still need to figure out what exactly that will involve.

**Comment by tbssajal** [comment]
I think in case of 100% slash, if we jail the validator until every existing delegator (non-self-bonded) unbonds from the validator, that will solve the issue.

**Comment by tbssajal** [comment]
> I think in case of 100% slash, if we jail the validator until every existing delegator (non-self-bonded) unbonds from the validator, that will solve the issue.

If we approach this way, then I suggest to add a new field `bool killed` in `struct Validator` and we don't add any new `ValidatorState` to keep the modification in logic simple. So when a validator is slashed 100%, `validator.state = ValidatorState.jailed` as previously done and we also set `validator.killed = true`. The we only need to modify the logic of `function activateValidator`. When activating the validator, we do an additional check: `if (validator.killed == true)` then `validator.liquidSupply == 0` has to be true to activate the validator. If we can activate him, then we also set `validator.killed = false`. This way, when a validator is killed, he is also in `jailed` state, so hopefully all other logic related to jailing will not require any modification.

**Comment by yazzaoui** [comment]
To go back to this one I think the most simple outcome after killing a validator (100% slash) is to disable it permanently. Don't see any benefit in practice to re-allow a 100% slashed validator to come back to life after a long period of unbonding as in practice this will probably never happen. Every delegator account need to unbond.

I would like as well to have a separate state for this concept instead of a flag... I agree the changes in the logic would be lessened with a flag but from a conceptual perspective really you are creating 3 more states :
active-not-killed/active-killed , disabled-not-killed/disabled-killed and jailed-killed/jailed-not-killed.

Comments?

**Comment by tbssajal** [comment]
Agreed with both points.

**Comment by lorenzo-dev1** [comment]
Yes I think we should go with the cleaner solution and add a state for killed. Also if a validator is slashed 100%, it means its correspective LNTN are now worthless? since they are not backed by NTNs no more

**Comment by cmjc** [comment]
> Yes I think we should go with the cleaner solution and add a state for killed. Also if a validator is slashed 100%, it means its correspective LNTN are now worthless? since they are not backed by NTNs no more

Yes. `_validator.bondedStake = 0` so there is no NTN to be redeemed either for _currently_ unbonding stake or for _future_ unbonding stake. There's zero value in paying for an unbond transaction in this scenario as there's nothing to redeem.

The only scenario I can think of to change this zero value scenario would be an 'insurance protocol' where an insurance fund pool was used to indemnify the slashed stake loss, the slashed stake was rebonded to the node. But that is academic as (a) you'd add a check to prevent bonding to a 'killed' state, (b) no such insurance protocol fund as of now!

**Comment by tbssajal** [comment]
>I'm struggling with how you can be in a state of active-killed

@cmjc I don't think it is possible to be in active-killed state, but what Youssef was trying to say is that by putting a `bool` field `killed`, we will create two options, `killed = true` and `killed = false` for each state, even though some states will not be reachable. But in terms of readability, this new `bool` field `killed` will create unnecessary options. So the suggestion of creating new state `killed` makes sense to me as well.

**Comment by tbssajal** [comment]
Resolved by #857
